File: julia/src/data_structures/doubly_linked_list.jl
"""
Doubly Linked List - Bidirectional linked list

Time Complexity:
- prepend!: O(1)
- append!: O(1)
- remove_first!: O(1)
- remove_last!: O(1)
- get: O(n)
"""

mutable struct DLLNode{T}
    data::T
    prev::Union{DLLNode{T}, Nothing}
    next::Union{DLLNode{T}, Nothing}
end

mutable struct DoublyLinkedList{T}
    head::Union{DLLNode{T}, Nothing}
    tail::Union{DLLNode{T}, Nothing}
    count::Int

    DoublyLinkedList{T}() where T = new{T}(nothing, nothing, 0)
end

DoublyLinkedList() = DoublyLinkedList{Any}()

"""
    prepend!(list::DoublyLinkedList{T}, item::T) where T

Add an item to the front.
"""
function prepend!(list::DoublyLinkedList{T}, item::T) where T
    node = DLLNode{T}(item, nothing, list.head)
    if !isnothing(list.head)
        list.head.prev = node
    end
    list.head = node
    if isnothing(list.tail)
        list.tail = node
    end
    list.count += 1
    list
end

"""
    append!(list::DoublyLinkedList{T}, item::T) where T

Add an item to the back.
"""
function Base.append!(list::DoublyLinkedList{T}, item::T) where T
    node = DLLNode{T}(item, list.tail, nothing)
    if !isnothing(list.tail)
        list.tail.next = node
    end
    list.tail = node
    if isnothing(list.head)
        list.head = node
    end
    list.count += 1
    list
end

"""
    get(list::DoublyLinkedList{T}, index::Int) where T

Get item at index (1-based).
"""
function get(list::DoublyLinkedList{T}, index::Int) where T
    if index < 1 || index > list.count
        throw(BoundsError(list, index))
    end
    # Start from closer end
    if index <= list.count ÷ 2
        current = list.head
        for _ in 1:(index-1)
            current = current.next
        end
    else
        current = list.tail
        for _ in 1:(list.count - index)
            current = current.prev
        end
    end
    current.data
end

Base.getindex(list::DoublyLinkedList, i::Int) = get(list, i)

"""
    set!(list::DoublyLinkedList{T}, index::Int, item::T) where T

Set item at index (1-based).
"""
function set!(list::DoublyLinkedList{T}, index::Int, item::T) where T
    if index < 1 || index > list.count
        throw(BoundsError(list, index))
    end
    if index <= list.count ÷ 2
        current = list.head
        for _ in 1:(index-1)
            current = current.next
        end
    else
        current = list.tail
        for _ in 1:(list.count - index)
            current = current.prev
        end
    end
    current.data = item
    list
end

"""
    remove_first!(list::DoublyLinkedList{T}) where T

Remove and return the first item.
"""
function remove_first!(list::DoublyLinkedList{T}) where T
    if isnothing(list.head)
        throw(ArgumentError("List is empty"))
    end
    result = list.head.data
    list.head = list.head.next
    if !isnothing(list.head)
        list.head.prev = nothing
    else
        list.tail = nothing
    end
    list.count -= 1
    result
end

"""
    remove_last!(list::DoublyLinkedList{T}) where T

Remove and return the last item.
"""
function remove_last!(list::DoublyLinkedList{T}) where T
    if isnothing(list.tail)
        throw(ArgumentError("List is empty"))
    end
    result = list.tail.data
    list.tail = list.tail.prev
    if !isnothing(list.tail)
        list.tail.next = nothing
    else
        list.head = nothing
    end
    list.count -= 1
    result
end

"""
    remove_at!(list::DoublyLinkedList{T}, index::Int) where T

Remove and return item at index (1-based).
"""
function remove_at!(list::DoublyLinkedList{T}, index::Int) where T
    if index < 1 || index > list.count
        throw(BoundsError(list, index))
    end
    if index == 1
        return remove_first!(list)
    end
    if index == list.count
        return remove_last!(list)
    end

    if index <= list.count ÷ 2
        current = list.head
        for _ in 1:(index-1)
            current = current.next
        end
    else
        current = list.tail
        for _ in 1:(list.count - index)
            current = current.prev
        end
    end
    result = current.data
    current.prev.next = current.next
    current.next.prev = current.prev
    list.count -= 1
    result
end

"""
    isempty(list::DoublyLinkedList)

Check if the list is empty.
"""
Base.isempty(list::DoublyLinkedList) = list.count == 0

"""
    length(list::DoublyLinkedList)

Return the number of items.
"""
Base.length(list::DoublyLinkedList) = list.count

"""
    size(list::DoublyLinkedList)

Return the number of items.
"""
Base.size(list::DoublyLinkedList) = list.count

"""
    empty!(list::DoublyLinkedList)

Remove all items.
"""
function Base.empty!(list::DoublyLinkedList)
    list.head = nothing
    list.tail = nothing
    list.count = 0
    list
end

"""
    collect(list::DoublyLinkedList{T}) where T

Convert to array.
"""
function Base.collect(list::DoublyLinkedList{T}) where T
    result = T[]
    current = list.head
    while !isnothing(current)
        push!(result, current.data)
        current = current.next
    end
    result
end

# Iterator interface
function Base.iterate(list::DoublyLinkedList, state=list.head)
    isnothing(state) ? nothing : (state.data, state.next)
end

File: lua/data_structures/doubly_linked_list.lua
-- Doubly Linked List - Bidirectional linked list
--
-- Time Complexity:
-- - prepend: O(1)
-- - append: O(1)
-- - removeFirst: O(1)
-- - removeLast: O(1)
-- - get: O(n)

local DoublyLinkedList = {}
DoublyLinkedList.__index = DoublyLinkedList

local function newNode(data)
    return { data = data, prev = nil, next = nil }
end

function DoublyLinkedList.new()
    local self = setmetatable({}, DoublyLinkedList)
    self.head = nil
    self.tail = nil
    self.count = 0
    return self
end

function DoublyLinkedList:prepend(item)
    local node = newNode(item)
    node.next = self.head
    if self.head ~= nil then
        self.head.prev = node
    end
    self.head = node
    if self.tail == nil then
        self.tail = node
    end
    self.count = self.count + 1
end

function DoublyLinkedList:append(item)
    local node = newNode(item)
    node.prev = self.tail
    if self.tail ~= nil then
        self.tail.next = node
    end
    self.tail = node
    if self.head == nil then
        self.head = node
    end
    self.count = self.count + 1
end

function DoublyLinkedList:get(index)
    if index < 1 or index > self.count then
        error("Index out of bounds")
    end
    local current
    if index <= self.count / 2 then
        current = self.head
        for _ = 1, index - 1 do
            current = current.next
        end
    else
        current = self.tail
        for _ = 1, self.count - index do
            current = current.prev
        end
    end
    return current.data
end

function DoublyLinkedList:set(index, item)
    if index < 1 or index > self.count then
        error("Index out of bounds")
    end
    local current
    if index <= self.count / 2 then
        current = self.head
        for _ = 1, index - 1 do
            current = current.next
        end
    else
        current = self.tail
        for _ = 1, self.count - index do
            current = current.prev
        end
    end
    current.data = item
end

function DoublyLinkedList:removeFirst()
    if self.head == nil then
        error("List is empty")
    end
    local item = self.head.data
    self.head = self.head.next
    if self.head ~= nil then
        self.head.prev = nil
    else
        self.tail = nil
    end
    self.count = self.count - 1
    return item
end

function DoublyLinkedList:removeLast()
    if self.tail == nil then
        error("List is empty")
    end
    local item = self.tail.data
    self.tail = self.tail.prev
    if self.tail ~= nil then
        self.tail.next = nil
    else
        self.head = nil
    end
    self.count = self.count - 1
    return item
end

function DoublyLinkedList:isEmpty()
    return self.count == 0
end

function DoublyLinkedList:size()
    return self.count
end

function DoublyLinkedList:clear()
    self.head = nil
    self.tail = nil
    self.count = 0
end

function DoublyLinkedList:toArray()
    local result = {}
    local current = self.head
    local index = 1
    while current ~= nil do
        result[index] = current.data
        current = current.next
        index = index + 1
    end
    return result
end

return DoublyLinkedList
